src/mono/browser/runtime/hybrid-globalization/calendar.ts

```diff
@@ -326,3 +329,37 @@ function getEraNames (date: Date, locale: string | undefined, calendarId: number
         };
     }
 }
+
+// wraps all substrings in the format in quotes, except for key words
+// transform e.g. "dddd, d MMMM yyyy г." into "dddd, d MMMM yyyy 'г'."
+function wrapSubstrings (str: string, locale: string | undefined) {
+    const words = str.split(" ");
+    // locales that write date nearly without spaces should not have format parts quoted - "ja", "zh"
+    // "ko" format parts should not be quoted but processing it would overcomplicate the logic
+    if (words.length <= 2 || locale?.startsWith("ko")) {
+        return str;
+    }
+
+    const keyWords = [MONTH_CODE, YEAR_CODE, DAY_CODE, WEEKDAY_CODE];
+    for (let i = 0; i < words.length; i++) {
+        if (!keyWords.includes(words[i].replace(",", "")) &&
```

Comment:
could all of the `replace()` be moved before the `str.split` ?

Reply:
yes

Reply:
Wait, why? It is more costly to do so

Reply:
okay, could we rewrite the whole loop + if into single compiled regex ?
`(MMMM|d|dddd)` for matching keywords and capture groups around them.
Compiled regex is fast, because it will typically only scan the input once. There is state machine/parser it will generate internally.

Reply:
This regex would be very complex and in case something changes in unicode standare, debugging would be a nightmare. It's not about just grepping "MMMM" etc. It's about recognizing substrings that should be in quotes and putting them into these quotes. There's a couple of edge cases, e.g.
- "yyyy. MMMM d., dddd" - we have to differentiate between "d.," that is typically a placeholder for day number in  "hu-HU" format from "d.," that is short from "diena" in "yyyy 'm'. MMMM d 'd'., dddd" - "lt-LT" format
- "dddd, yyyy. 'gada' d. MMMM" in "lv-LV" has "d." but it is a day placeholder, not a fixed expression.

Even if we made a compromise:
leaving the loop but using regex instead of long "if-case" inside of the loop then we have a problem with JS lacking support of lookbehinds in regular expressions. We need them because we want to clean the substrings that we got by splitting the format, from "," "." "\u05d1" and "\u060c" simultaneously not cleaning up ".,".

```
const words = str.split(/\s+/);
for (let i = 0; i < words.length; i++) {
        const cleanedWord = words[i].replace(/[.,\u060c\u05d1]/g, "");
        if (!keyWords.includes(cleanedWord)) {
               // ......wrap in quotes
       }
       return words.join(" ");
```

Reply:
Sorry, there is lookbehind:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Regular_expressions/Lookbehind_assertion

Reply:
we can drop it if it's complex rabbit hole. Nobody loves debugging regex.